sphinxcontrib/prettyspecialmethods.py

```diff
@@ -213,7 +215,9 @@ def show_special_methods(app, what, name, obj, skip, options):
 
 def setup(app):
     # type: (Sphinx) -> Dict[str, Any]
-    app.add_transform(PrettifySpecialMethods)
+    # Priority must be lower than Python domain meta filter to ensure it runs
+    # before meta info fields are removed.
```

Comment:
Does a similar reason force us to not use `SphinxTransform`, or did you just switch to `'object-description-transform'` for simplicity?

Reply:
Yes, this was the reason I switched away from `SphinxTransform`. I couldn't find any mechanism for getting an instance of `SphinxTransform` to run before `meta` was filtered out. Even setting the `default_priority` didn't make it run soon enough. It is entirely possible I just don't know what I'm doing, but I was pretty thorough in looking. I did quite a bit of hooking up events and debugging to see what order things happened in, and I looked through a decent bit of source code trying to find something. (Frankly, the Sphinx source code was too confusing for me to *find* where transforms are actually applied.) This event was always fired well before the transform was applied, and you can see [in the Python domain](https://github.com/sphinx-doc/sphinx/blob/master/sphinx/domains/python.py#L1298) that this event is where `meta` filtering happens. It did seem like a logical choice as well: we're applying a transform on method objects after it gets initially parsed.

Reply:
I've done some more research on this, and while I could be mistaken because it's difficult to piece together what is invoked where and in what order, my findings seem to confirm that there's not a better way to do this at present.

As noted before, `meta` info fields are cleaned on the [object-description-transform](https://www.sphinx-doc.org/en/master/extdev/appapi.html#event-object-description-transform) event. This event occurs when an `ObjectDescription` directive [is run](https://github.com/sphinx-doc/sphinx/blob/3.x/sphinx/directives/__init__.py#L211).

The documentation on Sphinx's [core events](https://www.sphinx-doc.org/en/master/extdev/appapi.html#sphinx-core-
events) *seems* to suggest that `SphinxTransform`s added to the application object are run much later than these directives. Specifically, they're run in the same phase as the `doctree-read`, which occurs after the document is parsed and directives are executed.

Maybe I should look at implementing a custom info field instead of using `meta`?

Reply:
@eric-wieser Sorry for all the messages.

Somewhere along the line in reading more, I stumbled across the [`pending` element](https://docutils.sourceforge.io/docs/howto/rst-directives.html#the-pending-element). I figured out how to use this [to implement a directive](https://github.com/johnpmcconnell/prettyspecialmethods/blob/feature/variation/variable-self-ref/custom-directive/sphinxcontrib/prettyspecialmethods.py) that leaves a sort of "marker" for the transform to process later. It looks like this in usage:

```
.. py:method:: __await__()

   .. selfref:: obj
```

I also [implemented a custom info field](https://github.com/johnpmcconnell/prettyspecialmethods/blob/feature/variation/variable-self-ref/custom-info-field/sphinxcontrib/prettyspecialmethods.py), which looks like this:

```
.. method:: __add__(other)
    Docstring
    :selfref: obj
```

Both of these implementations preserve the usage of `SphinxTransform`. I'm not sure which is more in line with normal docutils/Sphinx patterns. Both feel like they break the mold to me:

- `pending` was really designed to invoke its own transform, rather than act as a parameter for another one. While I *could* potentially implement a second transform that goes back and does even more to the signature after `PrettifySpecialMethods` does its work, I felt like that posed a much greater risk of some kind of mistake having an unintended side effect or some particular form of the signature ending up wrong. But I can go back and see about it if you really think I should.
- The info fields aren't designed to be parameters for some other transform either. They're supposed to specify additional information about the object being described, except for `meta` which is really only used by autodoc for filtering purposes.

It doesn't seem like augmenting an existing directive with additional transforms and parameters is something the system is really designed for, unless I'm missing something.

Anyway, please take a look at both of those implementations and let me know what approach you feel is best going forward.

Reply:
Thanks for the thorough investigation - I agree, the `meta` version seems simplest right now, but feel free to keep looking for  better approach if you want to.

Otherwise, I think modulo a little more thought about better words for `selfref`, this is good to merge.

Reply:
@eric-wieser The more I work with this and consider it, the more I think that a directive is a better choice than an info field, even a meta one. Using a directive lets us avoid hooking into events entirely, avoids potential conflicts with changes to Sphinx's Python domain, and is more semantically correct.

* Directives are the standard mechanism for *mark up*, and this parameter represents a kind of mark up that changes how the final output is rendered. Info fields provide additional descriptive information about the item being documented, and this is not additional descriptive information.
* The directive also lets us very cleanly segregate our code and elements from everyone else's, making it less likely that some other extension will accidentally process ours incorrectly.

In a more ideal world, a new option to `py:method` would be the *most* semantically correct, but I don't believe that extending it is a good idea even if it's possible. As far as I can tell, it is *not* a supported approach, meaning even a working solution is going to be hacky, fragile, and likely to conflict with other extensions trying to do similar tasks.

I suppose the one argument in favor of an info field is that it *looks* like an option to the `py:method` directive, but I'm not sure that's necessarily a good thing. It requires the user to place whitespace before it if writing reST directly, which is confusing if we want users to think of it 

Do you have any strong objection to using a directive instead?